shell/platform/android/io/flutter/plugin/editing/TextEditingDelta.java

```diff
@@ -0,0 +1,200 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugin.editing;
+
+import androidx.annotation.VisibleForTesting;
+import io.flutter.Log;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/// A representation of the change that occured to an editing state, along with the resulting
+/// composing and selection regions.
+public class TextEditingDelta {
+  private CharSequence oldText;
+  private CharSequence deltaText;
+  private CharSequence deltaType;
+  private int deltaStart;
+  private int deltaEnd;
+  private int newSelectionStart;
+  private int newSelectionEnd;
+  private int newComposingStart;
+  private int newComposingEnd;
+
+  private static final String TAG = "TextEditingDelta";
+
+  public TextEditingDelta(
```

Comment:
I was expecting that we can just extract the `tb{start, end}`, `start`, `end` and `tb` from the `replace` implementation and send them verbatim to the framework, and the framework will have to implement the same `replace` method to apply it to a `TextEditingValue`. Would that not work?

Reply:
That wouldn't work in some cases. Say the case where we are typing "world". 

Current state: "worl|"
soft keyboard input: "d"
replace method call: replace(0,4, world, 0, 5)

or for a deletion

Current state: "world|"
soft keyboard input: backspace
replace method call: replace(0,5, worl, 0, 4)

If typing at the end of the composing region the text in the composing region is replaced with the entire new composing text.

If we move the cursor inside the word currently being composed like below we get this call.
current state: "wo|ld"
soft keyboard input: "r"
replace method call: replace(2,2, r, 0,1)

Here we are actually given the insertion point and the character inserted, where when it is at the end of the composing region we must do some reasoning. 

On iOS we have something slightly different as well.

Current state: "world|"
soft keyboard input: backspace
replaceRangeWithLocal (4,5) with empty string ranges (0,0)

Because these values can vary a bit by platform for the same operation I think it is important to unify them into one format. I'm unsure if a replacement of "wor**ld**" with "worl" would have the same result as replacing "d" in "world" with an empty string regarding if ranges of "world" where bolded/styled. Would these ranges still have the information needed to contract/expand as expected?

Reply:
Unfortunately from my research of the API I don't think so. Even the `InputConnectionAdaptor` does not provide these granular differences between editing states. 

`InputConnectionAdaptor.commitText` and `InputConnectionAdaptor.setComposingText` both call their super class `BaseInputConnection` which calls `replaceText()` and routes that to `SpannableStringBuilder.replace()`. Only in some cases will the IME actually provide granular changes like in `InputConnectionAdaptor.deleteSurroundingText` which does call `SpannableStringBuilder.delete()` that is in the end also routed to `replace()`. In most cases for Android since the composing region is always active we are always replacing the entire composing region. 

Maybe I missed something but this is my conclusion after my research.

Reply:
I see. So `setComposingText` and `commit*` can delete/insert/replace, but they all only operate on a specific range.

Reply:
I'd prefer a bit more documentation in this class. If this were the entrypoint for someone learning about this feature, the documentation here is a bit lacking to properly introduce what this does and find the rest of the code.